File: Hash/src/main/java/Solution811.java

import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;

public class Solution811 {
    public List<String> subdomainVisits(String[] cpdomains) {
        HashMap<String, Integer> resMap = new HashMap<>();

        for (String cpdomain : cpdomains) {
            String[] cpdomainInfo = cpdomain.split(" ");
            Integer count = Integer.valueOf(cpdomainInfo[0]);

            // Java 的split的参数是 正则
            String[] frags = cpdomainInfo[1].split("\\.");

            String domain = "";
            for (int i = frags.length - 1; i >= 0; i--) {
                domain = frags[i] + (i == frags.length - 1 ? "" : ".") + domain;
                resMap.put(domain, resMap.getOrDefault(domain, 0) + count);
            }

        }

        List<String> ans = new ArrayList<>();

        for (String key : resMap.keySet()) {
            String res = resMap.get(key) + " " + key;
            ans.add(res);

        }

        return ans;

    }

    public static void main(String[] args) {
        Solution811 s = new Solution811();
        String[] cpdomains = { "900 google.mail.com", "50 yahoo.com", "1 intel.mail.com", "5 wiki.org" };

        List<String> res = s.subdomainVisits(cpdomains);

        System.out.println(res);
    }

}

File: Map/src/main/python/solution811.py
from typing import List


class Solution:
    def subdomainVisits(self, cpdomains: List[str]) -> List[str]:
        res_map = {}
        for one in cpdomains:
            one_l = one.split(" ")
            count = int(one_l[0])
            one_domain = one_l[1]

            all_domains = one_domain.split(".")

            if len(all_domains) == 2:
                all_domains = [one_domain, all_domains[-1]]
            if len(all_domains) == 3:
                all_domains = [one_domain, all_domains[-1], ".".join(all_domains[1:3])]

            for domain in all_domains:
                if res_map.get(domain):
                    res_map[domain] += count
                else:
                    res_map[domain] = count

        res_lst = []
        for k, v in res_map.items():
            res = " ".join([str(v), k])
            res_lst.append(res)

        return res_lst


s = Solution()
# cpdomains = ["9001 discuss.leetcode.com"]
cpdomains = ["900 google.mail.com", "50 yahoo.com", "1 intel.mail.com", "5 wiki.org"]
res = s.subdomainVisits(cpdomains)
print(res)
